src/radical/pilot/agent/executing/popen.py

```diff
@@ -190,6 +190,12 @@ def _handle_unit(self, cu):
             if cpt == 'MPI': launcher = self._mpi_launcher
             else           : launcher = self._task_launcher
 
+            if gpt == 'CUDA':
+                # This assumes single GPU device per unit. FIXME: generalize for 
+                # mulitple GPU devices
+                gpu_device = cu['slots']['nodes'][0][3][0][0]
+                cu['description']['environment']['CUDA_VISIBLE_DEVICES'] = gpu_device
```

Comment:
Mainly because I think of the scheduler as a component that only takes decisions where to place the units. The executor is then responsible to make sure that the environment is such so that the unit executes correctly.

Reply:
I do not think the ABDS launcher will need something like that. I agree that the ORTE and SSH executors will need it, but do we have any use cases that use those and need GPUs?

I am not sure how to test them. Let me take a look at that though

Reply:
> Mainly because I think of the scheduler as a component that only takes decisions where to place the units. The executor is then responsible to make sure that the environment is such so that the unit executes correctly.

Ack on that.

Reply:
> I do not think the ABDS launcher will need something like that. I agree that the ORTE and SSH executors will need it, but do we have any use cases that use those and need GPUs?
> 
> I am not sure how to test them. Let me take a look at that though

So what will the ABDS launcher do on `GPT=CUDA`?  

I don't really know about use cases - but I do know that the launchers should be semantically in sync.

Reply:
Nothing. It will launch a YARN executable. At that point YARN is responsible to schedule everything correctly in the Hadoop cluster.

Let me think of a way to include it to the other executors. The ABDS may be the only exception.

Reply:
> Nothing. It will launch a YARN executable. At that point YARN is responsible to schedule everything correctly in the Hadoop cluster.

How does Yarn know that this is supposed to land on a GPU?

Reply:
There is a [`XML`](https://hadoop.apache.org/docs/r3.1.0/hadoop-yarn/hadoop-yarn-site/UsingGpus.html) file that includes the types of resources YARN.